Issue with implementation_name env marker on py34

On Python 3.4 with minimum package levels, `make installtest` fails because Pip stumbles across the use of the implementation_name env marker in dev-requirements.txt and minimum-constraints.txt that was introduced with PR #2200 on the `typed-ast` package:

```
makefile: Running install tests
tests/installtest/test_install.sh dist/pywbem-1.0.0.dev1-py2.py3-none-any.whl dist/pywbem-1.0.0.dev1.tar.gz python
Preparing for the tests
Creating test directory: tests/installtest/../../tmp_installtest
Creating source archive unpack directory: tests/installtest/../../tmp_installtest/src_dist
Removing .egg file: tests/installtest/../../dist/pywbem*.egg
Testcase test1: Pip install from repo root directory: tests/installtest/../..
Saving current virtualenv: /home/travis/virtualenv/python3.4.8
Creating virtualenv: tests/installtest/../../tmp_installtest/virtualenvs/pywbem_test_test1
..using this Python command: /home/travis/virtualenv/python3.4.8/bin/python
..that has this Python version: Python 3.4.8
Activating virtualenv: tests/installtest/../../tmp_installtest/virtualenvs/pywbem_test_test1
Python command in this virtualenv: /home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/bin/python
Python version in this virtualenv: Python 3.4.8
Packages in this virtualenv before reinstalling base packages:
pip (8.0.0)
setuptools (19.4)
wheel (0.26.0)
Reinstalling setuptools with PACKAGE_LEVEL=minimum
Error: Command failed with rc=2: pip install setuptools -c /home/travis/build/pywbem/pywbem/minimum-constraints.txt, output follows:
Ignoring pip: markers "python_version == '3.6'" don't match your environment
Ignoring pip: markers "python_version == '3.7'" don't match your environment
Ignoring pip: markers "python_version >= '3.8'" don't match your environment
Ignoring setuptools: markers "python_version == '3.7'" don't match your environment
Ignoring setuptools: markers "python_version >= '3.8'" don't match your environment
Ignoring wheel: markers "python_version == '3.7'" don't match your environment
Ignoring wheel: markers "python_version >= '3.8'" don't match your environment
Ignoring PyYAML: markers "python_version == '2.7'" don't match your environment
Ignoring six: markers "python_version >= '3.8'" don't match your environment
Ignoring funcsigs: markers "python_version < '3.3'" don't match your environment
Ignoring pytest: markers "python_version >= '3.7'" don't match your environment
Ignoring colorama: markers "python_version >= '3.5'" don't match your environment
Ignoring virtualenv: markers "python_version >= '3.5' and python_version < '3.8'" don't match your environment
Ignoring virtualenv: markers "python_version >= '3.8'" don't match your environment
Ignoring pluggy: markers "python_version >= '3.7'" don't match your environment
Ignoring lxml: markers "python_version >= '3.8'" don't match your environment
Ignoring pylint: markers "python_version == '2.7'" don't match your environment
Ignoring pylint: markers "python_version >= '3.5'" don't match your environment
Ignoring astroid: markers "python_version == '2.7'" don't match your environment
Ignoring astroid: markers "python_version >= '3.5'" don't match your environment
Exception:
Traceback (most recent call last):
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/basecommand.py", line 209, in main
    status = self.run(options, args)
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/commands/install.py", line 287, in run
    wheel_cache
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/basecommand.py", line 264, in populate_requirement_set
    requirement_set.add_requirement(req)
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/req/req_set.py", line 224, in add_requirement
    if not install_req.match_markers():
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/req/req_install.py", line 799, in match_markers
    return markers_interpret(self.markers)
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/_vendor/distlib/markers.py", line 190, in interpret
    return Evaluator(execution_context).evaluate(marker.strip())
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/_vendor/distlib/markers.py", line 100, in evaluate
    return handler(node)
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/_vendor/distlib/markers.py", line 163, in do_expression
    return self.evaluate(node.body)
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/_vendor/distlib/markers.py", line 100, in evaluate
    return handler(node)
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/_vendor/distlib/markers.py", line 127, in do_boolop
    result = self.evaluate(n)
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/_vendor/distlib/markers.py", line 100, in evaluate
    return handler(node)
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/_vendor/distlib/markers.py", line 147, in do_compare
    lhs = self.evaluate(lhsnode)
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/_vendor/distlib/markers.py", line 100, in evaluate
    return handler(node)
  File "/home/travis/build/pywbem/pywbem/tmp_installtest/virtualenvs/pywbem_test_test1/lib/python3.4/site-packages/pip/_vendor/distlib/markers.py", line 174, in do_name
    raise SyntaxError('invalid expression: %s' % node.id)
  File "<string>", line None
SyntaxError: invalid expression: implementation_name
makefile:693: recipe for target 'installtest' failed
```

--- a/dev-requirements.txt
+++ b/dev-requirements.txt
@@ -54,7 +54,9 @@ lxml>=4.2.4; python_version >= '3.5' and python_version <= '3.7'
 lxml>=4.4.1; python_version >= '3.8'
 
 # Virtualenv
-virtualenv>=14.0.0; python_version < '3.5'
+# Virtualenv 20.0.19 does not install pip into the virtualenv on Python 3.4.
+# TODO: Remove the circumvention of pinning virtualenv once solved.
+virtualenv>=14.0.0,<20.0.19; python_version < '3.5'
 virtualenv>=16.1.0; python_version >= '3.5' and python_version < '3.8'
 virtualenv>=20.0.0; python_version >= '3.8'
 

--- a/dev-requirements.txt
+++ b/dev-requirements.txt
@@ -102,9 +102,9 @@ astroid>=2.3.3; python_version >= '3.5'
 # Python 3.4 support.
 # Requiring typed-ast>=1.4.0 for Python 3.8 since it addresses compile errors
 # with missing pgenheaders.h and duplicate definition of a struct.
-typed-ast>=1.3.0,<1.4.0; python_version == '3.4' and implementation_name == "cpython"
-typed-ast>=1.3.0; python_version >= '3.5' and python_version < '3.8' and implementation_name == "cpython"
-typed-ast>=1.4.0; python_version >= '3.8' and implementation_name == "cpython"
+typed-ast>=1.3.0,<1.4.0; python_version == '3.4'
+typed-ast>=1.3.0; python_version >= '3.5' and python_version < '3.8'
+typed-ast>=1.4.0; python_version >= '3.8'
 
 # Flake8 and dependents (no imports, invoked via flake8 script):
 flake8>=3.7.9

--- a/dev-requirements.txt
+++ b/dev-requirements.txt
@@ -54,9 +54,8 @@ lxml>=4.2.4; python_version >= '3.5' and python_version <= '3.7'
 lxml>=4.4.1; python_version >= '3.8'
 
 # Virtualenv
-# Virtualenv 20.0.19 does not install pip into the virtualenv on Python 3.4.
-# TODO: Remove the circumvention of pinning virtualenv once solved.
-virtualenv>=14.0.0,<20.0.19; python_version < '3.5'
+# Virtualenv 20.0.19 has an issue where it does not install pip on Python 3.4.
+virtualenv>=14.0.0,!=20.0.19; python_version < '3.5'
 virtualenv>=16.1.0; python_version >= '3.5' and python_version < '3.8'
 virtualenv>=20.0.0; python_version >= '3.8'
 

--- a/minimum-constraints.txt
+++ b/minimum-constraints.txt
@@ -162,8 +162,8 @@ pylint==2.4.4; python_version >= '3.5'
 astroid==1.4.9; python_version == '2.7'
 astroid==2.1.0; python_version == '3.4'
 astroid==2.3.3; python_version >= '3.5'
-typed-ast==1.3.0; python_version >= '3.4' and python_version < '3.8' and implementation_name == "cpython"
-typed-ast==1.4.0; python_version >= '3.8' and implementation_name == "cpython"
+typed-ast==1.3.0; python_version >= '3.4' and python_version < '3.8'
+typed-ast==1.4.0; python_version >= '3.8'
 
 # Flake8 and dependents (no imports, invoked via flake8 script):
 flake8==3.7.9

--- a/tests/installtest/test_install.sh
+++ b/tests/installtest/test_install.sh
@@ -173,11 +173,9 @@ function make_virtualenv()
     pip list --format=columns 2>/dev/null || pip list 2>/dev/null
   fi
 
-  # If pip is reinstalled (=downgraded) before reinstalling setuptools and wheel,
-  # we get permission errors on Travis when running with minimum package levels.
+  run "pip install pip $PIP_OPTS" "Reinstalling pip with PACKAGE_LEVEL=$PACKAGE_LEVEL"
   run "pip install setuptools $PIP_OPTS" "Reinstalling setuptools with PACKAGE_LEVEL=$PACKAGE_LEVEL"
   run "pip install wheel $PIP_OPTS" "Reinstalling wheel with PACKAGE_LEVEL=$PACKAGE_LEVEL"
-  run "pip install pip $PIP_OPTS" "Reinstalling pip with PACKAGE_LEVEL=$PACKAGE_LEVEL"
 
   verbose "Virtualenv before actual install test:"
   verbose "Pip version: $(pip --version 2>&1)"

--- a/tests/installtest/test_install.sh
+++ b/tests/installtest/test_install.sh
@@ -136,25 +136,37 @@ function make_virtualenv()
   envnamep="${ENVPREFIX}$envname"
   envdir=$VIRTUALENV_DIR/$envnamep
 
-  echo "Saving current virtualenv: $VIRTUAL_ENV"
+  echo "Saving location of current virtualenv: $VIRTUAL_ENV"
   VIRTUAL_ENV_SAVED=$VIRTUAL_ENV
 
   if [[ -d $envdir ]]; then
+    verbose "Removing new virtualenv: $envdir"
     remove_virtualenv $envname
   fi
 
   python_cmd_path=$(which $PYTHON_CMD)
-  verbose "Creating virtualenv: $envdir"
-  verbose "..using this Python command: $python_cmd_path"
-  verbose "..that has this Python version: $($PYTHON_CMD --version 2>&1)"
-  run "virtualenv -p $python_cmd_path $envdir"
+  verbose "Before creating virtualenv: $envdir"
+  verbose "Python version: $($PYTHON_CMD --version 2>&1) from $(which $PYTHON_CMD)"
+  verbose "Pip version: $(pip --version 2>&1)"
+  if [[ "$DEBUG" == "true" ]]; then
+    echo "Debug: Python module path:"
+    python -c "import sys, pprint; pprint.pprint(sys.path)"
+    echo "Debug: Packages:"
+    pip list --format=columns 2>/dev/null || pip list 2>/dev/null
+  fi
 
+  run "virtualenv -p $python_cmd_path $envdir" "Creating virtualenv: $envdir"
   run "source $envdir/bin/activate" "Activating virtualenv: $envdir"
-  verbose "Python command in this virtualenv: $(which python)"
-  verbose "Python version in this virtualenv: $(python --version 2>&1)"
 
-  verbose "Packages in this virtualenv before reinstalling base packages:"
-  pip list --format=columns 2>/dev/null || pip list 2>/dev/null
+  verbose "Virtualenv before reinstalling base packages:"
+  verbose "Python version: $(python --version 2>&1) from $(which python)"
+  verbose "Pip version: $(pip --version 2>&1)"
+  if [[ "$DEBUG" == "true" ]]; then
+    echo "Debug: Python module path:"
+    python -c "import sys, pprint; pprint.pprint(sys.path)"
+    echo "Debug: Packages:"
+    pip list --format=columns 2>/dev/null || pip list 2>/dev/null
+  fi
 
   # If pip is reinstalled (=downgraded) before reinstalling setuptools and wheel,
   # we get permission errors on Travis when running with minimum package levels.
@@ -162,7 +174,9 @@ function make_virtualenv()
   run "pip install wheel $PIP_OPTS" "Reinstalling wheel with PACKAGE_LEVEL=$PACKAGE_LEVEL"
   run "pip install pip $PIP_OPTS" "Reinstalling pip with PACKAGE_LEVEL=$PACKAGE_LEVEL"
 
-  verbose "Packages in this virtualenv before actual install test:"
+  verbose "Virtualenv before actual install test:"
+  verbose "Pip version: $(pip --version 2>&1)"
+  verbose "Packages:"
   pip list --format=columns 2>/dev/null || pip list 2>/dev/null
 }
 
@@ -213,6 +227,7 @@ function run()
     echo "Debug: running in this shell: $cmd"
     eval "$cmd"
     rc=$?
+    echo "Debug: command returns: rc=$rc"
   else
     eval "$cmd" >$CMD_LOG_FILE 2>&1
     rc=$?
@@ -312,20 +327,15 @@ function assert_import_ok()
 {
   local module
   module="$1"
-  if [[ "$DEBUG" == "true" ]]; then
-    echo "Debug: module=$module"
-    details=""
-    #details="; sys.path=$(python -c 'import sys; print(sys.path)')"
-  else
-    details=""
-  fi
-  assert_run_ok "python -c \"import ${module}\"" "Python module '${module}' cannot be imported$details"
+  verbose "Checking for successful import of module: $module"
+  assert_run_ok "python -c \"import ${module}\"" "Python module '${module}' cannot be imported"
 }
 
 function assert_import_fails()
 {
   local module
   module="$1"
+  verbose "Checking for failing import of module: $module"
   assert_run_fails "python -c \"import ${module}\"" "Python module '${module}' can be imported but should have failed"
 }
 

--- a/tests/installtest/test_install.sh
+++ b/tests/installtest/test_install.sh
@@ -155,7 +155,12 @@ function make_virtualenv()
     pip list --format=columns 2>/dev/null || pip list 2>/dev/null
   fi
 
-  run "virtualenv -p $python_cmd_path $envdir" "Creating virtualenv: $envdir"
+  if [[ "$DEBUG" == "true" ]]; then
+    virtualenv_debug_opts="-vvv"
+  else
+    virtualenv_debug_opts=""
+  fi
+  run "virtualenv -p $python_cmd_path $virtualenv_debug_opts $envdir" "Creating virtualenv: $envdir"
   run "source $envdir/bin/activate" "Activating virtualenv: $envdir"
 
   verbose "Virtualenv before reinstalling base packages:"

